src/mono/mono/component/marshal-ilgen.c

```diff
@@ -0,0 +1,2861 @@
+	LAST = 0xff
+};
+#undef OPDEF
+
+#define mono_mb_emit_jit_icall(mb, name) (cb_to_mono->mb_emit_icall_id ((mb), MONO_JIT_ICALL_ ## name))
+
+static GENERATE_GET_CLASS_WITH_CACHE (date_time, "System", "DateTime");
+static GENERATE_TRY_GET_CLASS_WITH_CACHE (icustom_marshaler, "System.Runtime.InteropServices", "ICustomMarshaler");
+
+static void emit_string_free_icall (MonoMethodBuilder *mb, MonoMarshalConv conv);
+
+static void mono_marshal_ilgen_legacy_init (void);
+
+static MonoMarshalILgenCallbacks ilgen_marshal_cb;
+
+static IlgenCallbacksToMono *cb_to_mono;
+
+static bool
+marshal_ilgen_available (void)
+{
+	return true;
+}
+
+static MonoComponentMarshalILgen component_func_table = {
+	{ MONO_COMPONENT_ITF_VERSION, &marshal_ilgen_available },
+	&mono_marshal_ilgen_init_internal,
+	&mono_emit_marshal_ilgen,
+	&mono_marshal_ilgen_install_callbacks_mono,
+#ifndef ENABLE_ILGEN
+	&mono_marshal_noilgen_init_heavyweight
+#endif
+}; 
+
+
+MonoComponentMarshalILgen*
+mono_component_marshal_ilgen_init (void) 
+{
+	return &component_func_table;
+}
+
+static gboolean ilgen_cb_inited = FALSE;
+static MonoMarshalILgenCallbacks ilgen_marshal_cb;
+
+void
+mono_install_marshal_callbacks_ilgen (MonoMarshalILgenCallbacks *cb)
+{
+	g_assert (!ilgen_cb_inited);
+	g_assert (cb->version == MONO_MARSHAL_CALLBACKS_VERSION);
+	memcpy (&ilgen_marshal_cb, cb, sizeof (MonoMarshalILgenCallbacks));
+	ilgen_cb_inited = TRUE;
+}
+
+void
+mono_marshal_ilgen_install_callbacks_mono (IlgenCallbacksToMono *callbacks)
+{
+	cb_to_mono = callbacks;
+}
+
+static void
+emit_struct_free (MonoMethodBuilder *mb, MonoClass *klass, int struct_var)
+{
+	/* Call DestroyStructure */
+	/* FIXME: Only do this if needed */
+	cb_to_mono->mb_emit_byte (mb, MONO_CUSTOM_PREFIX);
+	cb_to_mono->mb_emit_op (mb, CEE_MONO_CLASSCONST, klass);
+	cb_to_mono->mb_emit_ldloc (mb, struct_var);
+	mono_mb_emit_jit_icall (mb, mono_struct_delete_old);
+}
+
+static int
+emit_marshal_array_ilgen (EmitMarshalContext *m, int argnum, MonoType *t,
+					MonoMarshalSpec *spec,
+					int conv_arg, MonoType **conv_arg_type,
+					MarshalAction action)
+{
+	MonoMethodBuilder *mb = m->mb;
+	MonoClass *klass = mono_class_from_mono_type_internal (t);
+	MonoMarshalNative encoding;
+
+	encoding = cb_to_mono->get_string_encoding (m->piinfo, spec);
+	MonoType *int_type = cb_to_mono->get_int_type ();
+	MonoType *object_type = cb_to_mono->get_object_type ();
+
+	MonoClass *eklass = m_class_get_element_class (klass);
+
+	switch (action) {
+	case MARSHAL_ACTION_CONV_IN:
+		*conv_arg_type = object_type;
+		conv_arg = cb_to_mono->mb_add_local (mb, object_type);
+
+		if (m_class_is_blittable (eklass)) {
+			cb_to_mono->mb_emit_ldarg (mb, argnum);
+			if (m_type_is_byref (t))
+				cb_to_mono->mb_emit_byte (mb, CEE_LDIND_I);
+			cb_to_mono->mb_emit_icall_id (mb, cb_to_mono->conv_to_icall (MONO_MARSHAL_CONV_ARRAY_LPARRAY, NULL));
+			cb_to_mono->mb_emit_stloc (mb, conv_arg);
+		} else {
+#ifdef DISABLE_NONBLITTABLE
```

Comment:
As follow-up cleanup we should get rid of the `DISABLE_NONBLITTABLE` define.  I think this was a previous attempt at partitioning the marshaling code.  as far as I can tell we never define `DISABLE_NONBLITTABLE` (not even with the `ENABLE_MINIMAL` cmake variable)

Reply:
I will do this along with the PR to remove the code when `ENABLE_ILGEN` is not defined.